enhancement(Datepicker) date range

gselderslaghs opened the pull request:
## Proposed changes
**Implemented date range select functionality as per pull-request #360** 
- This functionality adds an end date field when option `dateRange: true` is provided
- Implemented calendar modal functionality to select an end date
- Implemented end field input change handling and modifiers to adjust the year, month and day selectors if a click or enter keypress interaction is executed
- Daterange classes added for in-range days
_styling of the in-range items needs to be addressed, propose create this in a new issue since the global layout of the datepicker is outdated: currently still follows the M2 guidelines_

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue).
- [X] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).

## Checklist:
- [X] I have read the **[CONTRIBUTING document](https://github.com/materializecss/materialize/blob/main/CONTRIBUTING.md)**.
- [X] My commit messages follows the [conventional commit format](https://github.com/materializecss/materialize/blob/main/CONTRIBUTING.md#submitting-your-pull-request)
- [ ] My change requires a change to the documentation, and updated it accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

wuda-io commented:
I will merge it in first, and then change the modal in #556 because date and timepicker depend on the modal component. We should change it to the Popover API since it is widely available in browsers now. https://caniuse.com/?search=popover

gselderslaghs commented:
As Ive been working on the #360 multi date select, I've found some additional issue's that I'm about to commit soon regarding date range select

gselderslaghs commented:
Ok, implemented additional refactoring based on my previous issues concerns, think this PR is ready for review

**Changelog as of the latest commits**
- Implemented multiple global parameter for conditional check on single day or multi day date selection, used in calendar day click event handler to decide if it's a single day click on the calendar
- Refactored handleCalendarClick, split up single date click and date range click for the sake of code readability and simplicity
- Implemented setDateFromInput for logical navigation within the calendar by mouse/keypress interaction on input change, fixes issue where calendar only navigated to latest changed field previously, eg. when start and end date where set by user input, clicking or keypressing on any of both fields resulted in only navigating to the latest changed field
- Updated JSDoc, removed unused parameter
- Deprecated _setToStartOfDay (function has no effect)

wuda-io reviewed (approved):
LGTM 👍
